dali/pipeline/data/views.h

```diff
@@ -0,0 +1,150 @@
+// Copyright (c) 2019, NVIDIA CORPORATION. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef  DALI_PIPELINE_DATA_VIEWS_H_
+#define  DALI_PIPELINE_DATA_VIEWS_H_
+
+#include "dali/kernels/tensor_view.h"
+#include "dali/kernels/backend_tags.h"
+#include "dali/pipeline/data/tensor_list.h"
+#include "dali/pipeline/data/tensor.h"
+
+namespace dali {
+namespace detail {
+
+template <typename Backend>
+struct storage_tag_map;
+
+template <>
+struct storage_tag_map<CPUBackend> {
+  using type = kernels::StorageCPU;
+};
+
+template <>
+struct storage_tag_map<GPUBackend> {
+  using type = kernels::StorageGPU;
+};
+
+template <typename Backend>
+using storage_tag_map_t = typename storage_tag_map<Backend>::type;
+
+}  // namespace detail
+
+template <int ndim, typename Backend>
+kernels::TensorListShape<ndim> list_shape(const TensorList<Backend> &tl) {
+  auto &tshape = tl.tensor_shape(0);
+  if (ndim != kernels::DynamicDimensions)
+    DALI_ENFORCE((int)tshape.size() == ndim, "Input has a wrong number of dimensions");
+  return kernels::convert_dim<ndim>(kernels::TensorListShape<>(tl.shape()));
+}
+
+template <int ndim, typename Backend>
+kernels::TensorShape<ndim> tensor_shape(const TensorList<Backend> &tl) {
+  auto &tshape = tl.tensor_shape(0);
+  DALI_ENFORCE(tl.IsDenseTensor(), "Uniform, dense tensor expected");
```

Comment:
How should I do if I have a non-dense tensor?

Reply:
It's impossible. I can add a doxygen description stating exactly that.